flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierUnaligner.java

```diff
@@ -0,0 +1,420 @@
+	 * not perform due to the max configured unaligned checkpoint size.
+	 */
+	@Override
+	public boolean isBlocked(int channelIndex) {
+		return false;
+	}
+
+	/**
+	 * We still need to trigger checkpoint while reading the first barrier from one channel, because this might happen
+	 * earlier than the previous async trigger via mailbox by netty thread. And the {@link AbstractInvokable} has the
+	 * deduplication logic to guarantee trigger checkpoint only once finally.
+	 *
+	 * <p>Note this is also suitable for the trigger case of local input channel.
+	 */
+	@Override
+	public boolean processBarrier(
+			CheckpointBarrier receivedBarrier,
+			int channelIndex,
+			long bufferedBytes) throws IOException {
+		long barrierId = receivedBarrier.getId();
+		if (currentCheckpointId < barrierId) {
+			currentCheckpointId = barrierId;
+			numBarrierConsumed = 0;
+			Arrays.fill(barrierConsumedChannels, false);
+		}
+		if (currentCheckpointId == barrierId) {
+			barrierConsumedChannels[channelIndex] = true;
+			numBarrierConsumed++;
+		}
+		// processBarrier is called from task thread and can actually happen before notifyBarrierReceived on empty
+		// buffer queues
+		// to avoid replicating any logic, we simply call notifyBarrierReceived here as well
+		threadSafeUnaligner.notifyBarrierReceived(receivedBarrier, channelInfos[channelIndex]);
+		return false;
+	}
+
+	@Override
+	public boolean processCancellationBarrier(CancelCheckpointMarker cancelBarrier) throws Exception {
+		final long barrierId = cancelBarrier.getCheckpointId();
+
+		if (numBarrierConsumed > 0) {
+			// this is only true if some alignment is in progress and nothing was canceled
+
+			if (barrierId == currentCheckpointId) {
+				// cancel this alignment
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("{}: Checkpoint {} canceled, aborting alignment.", taskName, barrierId);
+				}
+
+				releaseBlocksAndResetBarriers();
+				notifyAbortOnCancellationBarrier(barrierId);
+			}
+			else if (barrierId > currentCheckpointId) {
+				// we canceled the next which also cancels the current
+				LOG.warn("{}: Received cancellation barrier for checkpoint {} before completing current checkpoint {}. " +
+						"Skipping current checkpoint.",
+					taskName,
+					barrierId,
+					currentCheckpointId);
+
+				// this stops the current alignment
+				releaseBlocksAndResetBarriers();
+
+				// the next checkpoint starts as canceled
+				currentCheckpointId = barrierId;
+				threadSafeUnaligner.setCurrentCheckpointId(currentCheckpointId);
+				notifyAbortOnCancellationBarrier(barrierId);
+			}
+
+			// else: ignore trailing (cancellation) barrier from an earlier checkpoint (obsolete now)
+
+		}
+		else if (barrierId > currentCheckpointId) {
+			// first barrier of a new checkpoint is directly a cancellation
+
+			// by setting the currentCheckpointId to this checkpoint while keeping the numBarriers
+			// at zero means that no checkpoint barrier can start a new alignment
```

Comment:
This currently doesn't hold.

Reply:
You are right. That's part of the other ticket. I'm not sure what to do at this point though.
Should I remove this whole logic, such that the other ticket can add it?
Or should I adjust the comment?
In general, as long as `CancellationMarker` are not also overtaking, we always might start the checkpoint accidentally after all.